### PR TITLE
Use C# 10

### PR DESCRIPTION
**Pull request by drewnoakes**
- Upgrades the compiler we use, and increases the `LangVersion` property to 10.0
- Use C#10 constant string interpolation
- Use implicit usings, and remove redundant using directives

Previous attempts at this were blocked by infra work. @MiYanni unblocked that with #7974.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7975)

### CODE REVIEW

**Comment by AraHaan** [comment]
Nice work guys, also why couldn't this be updated to use language version ``preview`` instead so it does not have to ever manually need to be changed?

I been using language version preview for years without issues even on my stable packages on my code.

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/LanguageServices/WorkspaceProjectContextHostInstanceTests.cs [hunk 1]
```diff
@@ -13,8 +10,6 @@ namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
 {
     public class WorkspaceProjectContextHostInstanceTests
     {
-        private readonly CommandLineArgumentsSnapshot _emptyCommandLineArguments = new(ImmutableArray<string>.Empty, isChanged: false);
```
Comment (MiYanni) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/LanguageServices/WorkspaceProjectContextHostInstanceTests.cs:16]: Was this unused?

**Review by MiYanni — APPROVED** [review]
Looks good.

File: build/import/Packages.targets [hunk 2]
```diff
@@ -29,7 +29,7 @@
     <PackageReference Update="RoslynTools.SignTool"                                                   Version="$(RoslynToolsSignToolVersion)" />
     <PackageReference Update="XliffTasks"                                                             Version="1.0.0-beta.20574.1" />
     <PackageReference Update="Microsoft.DiaSymReader.Pdb2Pdb"                                         Version="$(MicrosoftDiaSymReaderPdb2PdbVersion)" />
-    <PackageReference Update="Microsoft.Net.Compilers.Toolset"                                        Version="3.11.0" />
+    <PackageReference Update="Microsoft.Net.Compilers.Toolset"                                        Version="4.2.0-1.final" />
```
Comment (AraHaan) [inline build/import/Packages.targets:32]: Cant you use roslyn 4.1 stable for C# 10 as well?

Comment (drewnoakes) [reply build/import/Packages.targets:32]: Yes you can.

I would like to propose setting `LangVersion` to `preview` in a future PR that also uses the new string syntax. Some of our unit tests could be tidied up.

Also we try to dogfood previews for extra testing.

Comment (MiYanni) [reply build/import/Packages.targets:32]: Maybe get these changes in first and try `preview` in another PR with the unit test cleanup?

Comment (drewnoakes) [reply build/import/Packages.targets:32]: That's what I was thinking too.

**Pull request merged**